<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introduction" data-toc-modified-id="Introduction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introduction</a></span></li><li><span><a href="#Setup" data-toc-modified-id="Setup-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Setup</a></span><ul class="toc-item"><li><span><a href="#Setup---Debug" data-toc-modified-id="Setup---Debug-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Setup - Debug</a></span></li><li><span><a href="#Setup---Imports" data-toc-modified-id="Setup---Imports-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Setup - Imports</a></span></li><li><span><a href="#Setup---logging" data-toc-modified-id="Setup---logging-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Setup - logging</a></span></li><li><span><a href="#Setup---virtualenv-jupyter-kernel" data-toc-modified-id="Setup---virtualenv-jupyter-kernel-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Setup - virtualenv jupyter kernel</a></span></li><li><span><a href="#Setup---Initialize-Django" data-toc-modified-id="Setup---Initialize-Django-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Setup - Initialize Django</a></span></li></ul></li><li><span><a href="#Find-articles-to-be-coded" data-toc-modified-id="Find-articles-to-be-coded-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Find articles to be coded</a></span><ul class="toc-item"><li><span><a href="#which-articles-have-already-been-coded?" data-toc-modified-id="which-articles-have-already-been-coded?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>which articles have already been coded?</a></span><ul class="toc-item"><li><span><a href="#Tag-the-coded-articles" data-toc-modified-id="Tag-the-coded-articles-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Tag the coded articles</a></span></li><li><span><a href="#Profile-the-coded-articles" data-toc-modified-id="Profile-the-coded-articles-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Profile the coded articles</a></span></li></ul></li><li><span><a href="#tag-all-local-news" data-toc-modified-id="tag-all-local-news-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>tag all local news</a></span><ul class="toc-item"><li><span><a href="#TODO" data-toc-modified-id="TODO-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>TODO</a></span><ul class="toc-item"><li><span><a href="#DONE" data-toc-modified-id="DONE-3.2.1.1"><span class="toc-item-num">3.2.1.1&nbsp;&nbsp;</span>DONE</a></span></li></ul></li><li><span><a href="#Grand-Rapids-Press-local-news" data-toc-modified-id="Grand-Rapids-Press-local-news-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Grand Rapids Press local news</a></span></li><li><span><a href="#Detroit-News-local-news" data-toc-modified-id="Detroit-News-local-news-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Detroit News local news</a></span></li></ul></li></ul></li><li><span><a href="#Code-Articles" data-toc-modified-id="Code-Articles-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Code Articles</a></span></li><li><span><a href="#TODO" data-toc-modified-id="TODO-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>TODO</a></span></li></ul></div>

# Introduction

- Back to [Table of Contents](#Table-of-Contents)

This is a notebook that expands on the OpenCalais code in the file `article_coding.py`, also in this folder.  It includes more sections on selecting publications you want to submit to OpenCalais as an example.  It is intended to be copied and re-used.

# Setup

- Back to [Table of Contents](#Table-of-Contents)

## Setup - Debug

- Back to [Table of Contents](#Table-of-Contents)

In [1]:
debug_flag = False

## Setup - Imports

- Back to [Table of Contents](#Table-of-Contents)

In [ ]:
import datetime
from django.db.models import Avg, Max, Min
import logging
import six

## Setup - logging

- Back to [Table of Contents](#Table-of-Contents)

configure logging for this notebook's kernel (If you do not run this cell, you'll get the django application's logging configuration.

In [ ]:
logging.basicConfig(
    level = logging.DEBUG,
    format = '%(asctime)s - %(levelname)s - %(name)s - %(message)s',
    filename = '/home/jonathanmorgan/logs/django-research.log',
    filemode = 'w' # set to 'a' if you want to append, rather than overwrite each time.
)

## Setup - virtualenv jupyter kernel

- Back to [Table of Contents](#Table-of-Contents)

If you are using a virtualenv, make sure that you:

- have installed your virtualenv as a kernel.
- choose the kernel for your virtualenv as the kernel for your notebook (Kernel --> Change kernel).

Since I use a virtualenv, need to get that activated somehow inside this notebook.  One option is to run `../dev/wsgi.py` in this notebook, to configure the python environment manually as if you had activated the `sourcenet` virtualenv.  To do this, you'd make a code cell that contains:

    %run ../dev/wsgi.py
    
This is sketchy, however, because of the changes it makes to your Python environment within the context of whatever your current kernel is.  I'd worry about collisions with the actual Python 3 kernel.  Better, one can install their virtualenv as a separate kernel.  Steps:

- activate your virtualenv:

        workon research

- in your virtualenv, install the package `ipykernel`.

        pip install ipykernel

- use the ipykernel python program to install the current environment as a kernel:

        python -m ipykernel install --user --name <env_name> --display-name "<display_name>"
        
    `sourcenet` example:
    
        python -m ipykernel install --user --name sourcenet --display-name "research (Python 3)"
        
More details: [http://ipython.readthedocs.io/en/stable/install/kernel_install.html](http://ipython.readthedocs.io/en/stable/install/kernel_install.html)

## Setup - Initialize Django

- Back to [Table of Contents](#Table-of-Contents)

First, initialize my dev django project, so I can run code in this notebook that references my django models and can talk to the database using my project's settings.

In [3]:
# init django
django_init_folder = "/home/jonathanmorgan/work/django/research/work/phd_work"
django_init_path = "django_init.py"
if( ( django_init_folder is not None ) and ( django_init_folder != "" ) ):
    
    # add folder to front of path.
    django_init_path = "{}/{}".format( django_init_folder, django_init_path )
    
#-- END check to see if django_init folder. --#

In [4]:
%run $django_init_path

django initialized at 2019-08-01 02:27:25.965530


In [5]:
# context_text imports
from context_text.article_coding.article_coding import ArticleCoder
from context_text.article_coding.article_coding import ArticleCoding
from context_text.article_coding.open_calais_v2.open_calais_v2_article_coder import OpenCalaisV2ArticleCoder
from context_text.collectors.newsbank.newspapers.GRPB import GRPB
from context_text.collectors.newsbank.newspapers.DTNB import DTNB
from context_text.models import Article
from context_text.models import Newspaper
from context_text.shared.context_text_base import ContextTextBase

# Find articles to be coded

- Back to [Table of Contents](#Table-of-Contents)

Tag all locally implemented hard news articles in database and all that have already been coded using Open Calais V2, then work through using OpenCalais to code all local hard news that hasn't alredy been coded, starting with those proximal to the coding sample for methods paper.

## which articles have already been coded?

- Back to [Table of Contents](#Table-of-Contents)

More precisely, find all articles that have Article_Data coded by the automated coder with type "OpenCalais_REST_API_v2" and tag the articles as "coded-open_calais_v2" or something like that.

Then, for articles without that tag, use our criteria for local hard news to filter out and tag publications in the year before and after the month used to evaluate the automated coder, in both the Grand Rapids Press and the Detroit News, so I can look at longer time frames, then code all articles currently in database.

Eventually, then, we'll code and examine before and after layoffs.

In [7]:
# look for publications that have article data:
# - coded by automated coder
# - with coder type of "OpenCalais_REST_API_v2"

# get automated coder
automated_coder_user = ArticleCoder.get_automated_coding_user()

print( "{} - Loaded automated user: {}, id = {}".format( datetime.datetime.now(), automated_coder_user, automated_coder_user.id ) )

2019-07-31 19:39:36.556744 - Loaded automated user: automated, id = 2


In [8]:
# try aggregates
article_qs = Article.objects.all()
pub_date_info = article_qs.aggregate( Max( 'pub_date' ), Min( 'pub_date' ) )
print( pub_date_info )

{'pub_date__max': datetime.date(2010, 11, 30), 'pub_date__min': datetime.date(2005, 1, 1)}


In [9]:
# find articles with Article_Data created by the automated user...
article_qs = Article.objects.filter( article_data__coder = automated_coder_user )

# ...and specifically coded using OpenCalais V2...
article_qs = article_qs.filter( article_data__coder_type = OpenCalaisV2ArticleCoder.CONFIG_APPLICATION )

# ...and finally, we just want the distinct articles by ID.
article_qs = article_qs.order_by( "id" ).distinct( "id" )

# count?
article_count = article_qs.count()
print( "Found {} articles".format( article_count ) )

Found 579 articles


### Tag the coded articles

- Back to [Table of Contents](#Table-of-Contents)

Removing duplicates present from joining with Article_Data yields 579 articles that were coded by the automated coder.

Tag all the coded articles with `OpenCalaisV2ArticleCoder.TAG_CODED_BY_ME`.

In [10]:
# declare variables
current_article = None
tag_name_list = None
article_count = None
untagged_count = None
already_tagged_count = None
newly_tagged_count = None
count_sum = None
do_add_tag = False

# init
do_add_tag = True

# get article_count
article_count = article_qs.count()

# loop over articles.
untagged_count = 0
already_tagged_count = 0
newly_tagged_count = 0
for current_article in article_qs:
    
    # get list of tags for this publication
    tag_name_list = current_article.tags.names()
    
    # is the coded tag in the list?
    if ( OpenCalaisV2ArticleCoder.TAG_CODED_BY_ME not in tag_name_list ):
        
        # are we adding tag?
        if ( do_add_tag == True ):

            # add tag.
            current_article.tags.add( OpenCalaisV2ArticleCoder.TAG_CODED_BY_ME )
            newly_tagged_count += 1
            
        else:

            # for now, increment untagged count
            untagged_count += 1
            
        #-- END check to see if we are adding tag. --#
        
    else:
        
        # already tagged
        already_tagged_count += 1
        
    #-- END check to see if coded tag is set --#
    
#-- END loop over articles. --#

print( "Article counts:" )
print( "- total articles: {}".format( article_count ) )
print( "- untagged articles: {}".format( untagged_count ) )
print( "- already tagged: {}".format( already_tagged_count ) )
print( "- newly tagged: {}".format( newly_tagged_count ) )
count_sum = untagged_count + already_tagged_count + newly_tagged_count
print( "- count sum: {}".format( count_sum ) )

Article counts:
- total articles: 579
- untagged articles: 0
- already tagged: 579
- newly tagged: 0
- count sum: 579


### Profile the coded articles

- Back to [Table of Contents](#Table-of-Contents)

Look at range of pub dates.

In [6]:
tags_in_list = []
tags_in_list.append( OpenCalaisV2ArticleCoder.TAG_CODED_BY_ME )
article_qs = Article.objects.filter( tags__name__in = tags_in_list )
print( "Matching article count: {}".format( article_qs.count() ) )

Matching article count: 589


- Original: 579
- after coding 10: 589 (tag is being set correctly bu Open Calais V2 coder)

In [12]:
# profile these publications
min_pub_date = None
max_pub_date = None
current_pub_date = None
pub_date_count = None
date_to_count_map = {}
date_to_articles_map = {}
pub_date_article_dict = None

# try aggregates
pub_date_info = article_qs.aggregate( Max( 'pub_date' ), Min( 'pub_date' ) )
print( pub_date_info )

# counts of pubs by date
for current_article in article_qs:
    
    # get pub_date
    current_pub_date = current_article.pub_date
    current_article_id = current_article.id
    
    # get count, increment, and store.
    pub_date_count = date_to_count_map.get( current_pub_date, 0 )
    pub_date_count += 1
    date_to_count_map[ current_pub_date ] = pub_date_count
    
    # also, store up ids and instances
    
    # get dict of article ids to article instances for date
    pub_date_article_dict = date_to_articles_map.get( current_pub_date, {} )
    
    # article already there?
    if ( current_article_id not in pub_date_article_dict ):
        
        # no - add it.
        pub_date_article_dict[ current_article_id ] = current_article
        
    #-- END check to see if article already there.
    
    # put dict back.
    date_to_articles_map[ current_pub_date ] = pub_date_article_dict
    
#-- END loop over articles. --#

# output dates and counts.

# get list of keys from map
keys_list = list( six.viewkeys( date_to_count_map ) )
keys_list.sort()
for current_pub_date in keys_list:
    
    # get count
    pub_date_count = date_to_count_map.get( current_pub_date, 0 )
    print( "- {} ( {} ) count: {}".format( current_pub_date, type( current_pub_date ), pub_date_count ) )
    
#-- END loop over dates --#

{'pub_date__max': datetime.date(2010, 7, 31), 'pub_date__min': datetime.date(2005, 1, 7)}
- 2005-01-07 ( <class 'datetime.date'> ) count: 1
- 2005-01-21 ( <class 'datetime.date'> ) count: 1
- 2005-02-07 ( <class 'datetime.date'> ) count: 1
- 2005-02-10 ( <class 'datetime.date'> ) count: 1
- 2005-05-10 ( <class 'datetime.date'> ) count: 1
- 2005-05-16 ( <class 'datetime.date'> ) count: 1
- 2005-06-06 ( <class 'datetime.date'> ) count: 1
- 2005-06-22 ( <class 'datetime.date'> ) count: 1
- 2005-07-02 ( <class 'datetime.date'> ) count: 1
- 2005-07-05 ( <class 'datetime.date'> ) count: 2
- 2005-07-07 ( <class 'datetime.date'> ) count: 1
- 2005-08-22 ( <class 'datetime.date'> ) count: 1
- 2005-09-04 ( <class 'datetime.date'> ) count: 1
- 2005-09-11 ( <class 'datetime.date'> ) count: 1
- 2005-09-25 ( <class 'datetime.date'> ) count: 1
- 2005-10-04 ( <class 'datetime.date'> ) count: 1
- 2005-10-20 ( <class 'datetime.date'> ) count: 1
- 2005-10-21 ( <class 'datetime.date'> ) count: 1
- 2005-10-

In [13]:
# look at the 2010-07-31 date
pub_date = datetime.datetime.strptime( "2010-07-31", "%Y-%m-%d" ).date()
articles_for_date = date_to_articles_map.get( pub_date, {} )
print( articles_for_date )

# get the article and look at its tags.
article_instance = articles_for_date.get( 6065 )
print( article_instance.tags.all() )

# loop over associated Article_Data instances.
for article_data in article_instance.article_data_set.all():
    
    print( article_data )
    
#-- END loop over associated Article_Data instances --#

{6065: <Article: 6065 - Jul 31, 2010, City and Region ( A6 ), UID: 1315C0760F2D0668 - Local ArtPeers registration opens ( Grand Rapids Press, The )>}
<QuerySet [<Tag: prelim_reliability_test>, <Tag: prelim_reliability_combined>, <Tag: coded-OpenCalaisV2ArticleCoder>]>
2180 - minnesota1 - no coder_type -- Article: 6065 - Jul 31, 2010, City and Region ( A6 ), UID: 1315C0760F2D0668 - Local ArtPeers registration opens ( Grand Rapids Press, The )
2200 - minnesota2 - no coder_type -- Article: 6065 - Jul 31, 2010, City and Region ( A6 ), UID: 1315C0760F2D0668 - Local ArtPeers registration opens ( Grand Rapids Press, The )
2281 - minnesota3 - no coder_type -- Article: 6065 - Jul 31, 2010, City and Region ( A6 ), UID: 1315C0760F2D0668 - Local ArtPeers registration opens ( Grand Rapids Press, The )
2969 - automated ( ADCT: OpenCalais_REST_API_v2 )  -- Article: 6065 - Jul 31, 2010, City and Region ( A6 ), UID: 1315C0760F2D0668 - Local ArtPeers registration opens ( Grand Rapids Press, The )


## tag all local news

- Back to [Table of Contents](#Table-of-Contents)

Definition of local hard news by in-house implementor for Grand Rapids Press and Detroit News follow.  For each, tag all articles in database that match as "local_hard_news".

### TODO

- Back to [Table of Contents](#Table-of-Contents)

TODO:

- make class for GRPB at NewsBank.

    - also, pull the things that are newspaper specific out of ArticleCoder.py and into the GRPB.py class.

- refine "local news" and "locally created" regular expressions for Grand Rapids Press based on contents of `author_string` and `author_affiliation`.
- do the same for TDN.
- then, use the updated classes and definitions below to flag all local hard news in database for each publication.

#### DONE

- Back to [Table of Contents](#Table-of-Contents)

DONE:

- abstract out shared stuff from GRPB.py and DTNB.py into abstract parent class context_text/collectors/newsbank/newspapers/newsbank_newspaper.py

    - update DTNB.py to use the parent class.
    
- make class for GRPB at NewsBank.

    - context_text/collectors/newsbank/newspapers/GRPB.py

### Grand Rapids Press local news

- Back to [Table of Contents](#Table-of-Contents)

Grand Rapids Press local hard news:

- `context_text/examples/articles/articles-GRP-local_news.py`
- local hard news sections (stored in `Article.GRP_NEWS_SECTION_NAME_LIST`):

    - "Business"
    - "City and Region"
    - "Front Page"
    - "Lakeshore"
    - "Religion"
    - "Special"
    - "State"

- in-house implementor (based on byline patterns, stored in `sourcenet.models.Article.Q_GRP_IN_HOUSE_AUTHOR`):

    - Byline ends in "/ THE GRAND RAPIDS PRESS", ignore case.

        - `Q( author_varchar__iregex = r'.* */ *THE GRAND RAPIDS PRESS$'`

    - Byline ends in "/ PRESS * EDITOR", ignore case.

        - `Q( author_varchar__iregex = r'.* */ *PRESS .* EDITOR$' )`

    - Byline ends in "/ GRAND RAPIDS PRESS * BUREAU", ignore case.

        - `Q( author_varchar__iregex = r'.* */ *GRAND RAPIDS PRESS .* BUREAU$' )`

    - Byline ends in "/ SPECIAL TO THE PRESS", ignore case.

        - `Q( author_varchar__iregex = r'.* */ *SPECIAL TO THE PRESS$' )`
        
- can also exclude columns (I will not):

        grp_article_qs = grp_article_qs.exclude( index_terms__icontains = "Column" )

Need to work to further refine this.

Looking at affiliation strings:

    SELECT author_affiliation, COUNT( author_affiliation ) as affiliation_count
    FROM context_text_article
    WHERE newspaper_id = 1
    GROUP BY author_affiliation
    ORDER BY COUNT( author_affiliation ) DESC;
    
And at author strings for collective bylines:

    SELECT author_string, COUNT( author_string ) as author_count
    FROM context_text_article
    WHERE newspaper_id = 1
    GROUP BY author_string
    ORDER BY COUNT( author_string ) DESC
    LIMIT 10;


In [16]:
# filter queryset to just locally created Grand Rapids Press (GRP) articles.
# imports
from context_text.models import Article
from context_text.models import Newspaper
from context_text.shared.context_text_base import ContextTextBase
from context_text.collectors.newsbank.newspapers.GRPB import GRPB

# declare variables - Grand Rapids Press
do_apply_tag = False
tag_to_apply = None
grp_local_news_sections = []
grp_newspaper = None
grp_article_qs = None
article_count = -1

# declare variables - filtering
include_opinion_columns = True
tags_in_list = []
tags_not_in_list = []
filter_out_prelim_tags = False
random_count = -1

# declare variables - make list of article IDs from QS.
article_id_list = []
article_counter = -1
current_article = None
article_tag_name_list = None
article_update_counter = -1

# ==> configure

# configure - size of random sample we want
#random_count = 60

# configure - also, apply tag?
do_apply_tag = True
tag_to_apply = ContextTextBase.TAG_LOCAL_HARD_NEWS

# set up "local, regional and state news" sections
grp_local_news_sections = GRPB.LOCAL_NEWS_SECTION_NAME_LIST

# Grand Rapids Press
# get newspaper instance for GRP.
grp_newspaper = Newspaper.objects.get( id = GRPB.NEWSPAPER_ID )

# start with all articles
#grp_article_qs = Article.objects.all()

# ==> filter to newspaper, local news section list, and in-house reporters.

# ----> manually

# now, need to find local news articles to test on.
#grp_article_qs = grp_article_qs.filter( newspaper = grp_newspaper )

# only the locally implemented sections
#grp_article_qs = grp_article_qs.filter( section__in = grp_local_news_sections )

# and, with an in-house author
#grp_article_qs = grp_article_qs.filter( Article.Q_GRP_IN_HOUSE_AUTHOR )

#print( "manual filter count: {}".format( grp_article_qs.count() ) )

# ----> using Article.filter_articles()
grp_article_qs = Article.filter_articles( qs_IN = grp_article_qs,
                                          newspaper = grp_newspaper,
                                          section_name_list = grp_local_news_sections,
                                          custom_article_q = GRPB.Q_IN_HOUSE_AUTHOR )

print( "Article.filter_articles count: {}".format( grp_article_qs.count() ) )

# and include opinion columns?
if ( include_opinion_columns == False ):
    
    # do not include columns
    grp_article_qs = grp_article_qs.exclude( index_terms__icontains = "Column" )
    
#-- END check to see if we include columns. --#

'''
# filter to newspaper, section list, and in-house reporters.
grp_article_qs = Article.filter_articles( qs_IN = grp_article_qs,
                                          start_date = "2009-12-01",
                                          end_date = "2009-12-31",
                                          newspaper = grp_newspaper,
                                          section_name_list = grp_local_news_sections,
                                          custom_article_q = Article.Q_GRP_IN_HOUSE_AUTHOR )
'''

# how many is that?
article_count = grp_article_qs.count()

print( "Article count before filtering on tags: " + str( article_count ) )

# ==> tags

# tags to exclude
tags_not_in_list = []

# Example: prelim-related tags
#tags_not_in_list.append( "prelim_reliability" )
#tags_not_in_list.append( "prelim_network" ]
#tags_not_in_list.append( "minnesota1-20160328" )
#tags_not_in_list.append( "minnesota2-20160328" )

# for later - exclude articles already coded.
#tags_not_in_list.append( OpenCalaisV2ArticleCoder.TAG_CODED_BY_ME )

# exclude any already tagged with tag_to_apply
tags_not_in_list.append( tag_to_apply )

if ( ( tags_not_in_list is not None ) and ( len( tags_not_in_list ) > 0 ) ):

    # exclude those in a list
    print( "filtering out articles with tags: " + str( tags_not_in_list ) )
    grp_article_qs = grp_article_qs.exclude( tags__name__in = tags_not_in_list )

#-- END check to see if we have a specific list of tags we want to exclude --#

# include only those with certain tags.
tags_in_list = []

# Examples

# Examples: prelim-related tags
#tags_in_list.append( "prelim_unit_test_001" )
#tags_in_list.append( "prelim_unit_test_002" )
#tags_in_list.append( "prelim_unit_test_003" )
#tags_in_list.append( "prelim_unit_test_004" )
#tags_in_list.append( "prelim_unit_test_005" )
#tags_in_list.append( "prelim_unit_test_006" )
#tags_in_list.append( "prelim_unit_test_007" )

# Example: grp_month
#tags_in_list.append( "grp_month" )

if ( ( tags_in_list is not None ) and ( len( tags_in_list ) > 0 ) ):

    # filter
    print( "filtering to just articles with tags: " + str( tags_in_list ) )
    grp_article_qs = grp_article_qs.filter( tags__name__in = tags_in_list )
    
#-- END check to see if we have a specific list of tags we want to include --#

# filter out "*prelim*" tags?
#filter_out_prelim_tags = True
if ( filter_out_prelim_tags == True ):

    # ifilter out all articles with any tag whose name contains "prelim".
    print( "filtering out articles with tags that contain \"prelim\"" )
    grp_article_qs = grp_article_qs.exclude( tags__name__icontains = "prelim" )
    
#-- END check to see if we filter out "prelim_*" tags --#

# how many is that?
article_count = grp_article_qs.count()

print( "Article count after tag filtering: " + str( article_count ) )

# do we want a random sample?
if ( random_count > 0 ):

    # to get random, order them by "?", then use slicing to retrieve requested
    #     number.
    grp_article_qs = grp_article_qs.order_by( "?" )[ : random_count ]
    
#-- END check to see if we want random sample --#

# this is a nice algorithm, also:
# - http://www.titov.net/2005/09/21/do-not-use-order-by-rand-or-how-to-get-random-rows-from-table/

# make ID list, tag articles if configured to.
article_id_list = []
article_counter = 0
article_update_counter = 0
for current_article in grp_article_qs:

    # increment article_counter
    article_counter += 1

    # add IDs to article_id_list
    article_id_list.append( str( current_article.id ) )
    
    # apply a tag while we are at it?
    if ( ( do_apply_tag == True ) and ( tag_to_apply is not None ) and ( tag_to_apply != "" ) ):
    
        # yes, please.  Tag already present?
        article_tag_name_list = current_article.tags.names()
        if ( tag_to_apply not in article_tag_name_list ):

            # Add tag.
            current_article.tags.add( tag_to_apply )
            
            # increment counter
            article_update_counter += 1
            
        #-- END check to see if tag already present. --#
        
    #-- END check to see if we apply tag. --#

    # output the tags.
    if ( debug_flag == True ):
        print( "- Tags for article " + str( current_article.id ) + " : " + str( current_article.tags.all() ) )
    #-- END DEBUG --#

#-- END loop over articles --#

# output the list.
print( "grp_article_qs count: {}".format( grp_article_qs.count() ) )
print( "Found " + str( article_counter ) + " articles ( " + str( article_count ) + " )." )
print( "- Updated {} articles to add tag {}.".format( article_update_counter, tag_to_apply ) )
if ( debug_flag == True ):
    print( "List of " + str( len( article_id_list ) ) + " local GRP staff article IDs: " + ", ".join( article_id_list ) )
#-- END DEBUG --#


Article.filter_articles count: 43816
Article count before filtering on tags: 43816
filtering out articles with tags: ['local_hard_news']
Article count after tag filtering: 0
grp_article_qs count: 0
Found 0 articles ( 0 ).
- Updated 0 articles to add tag local_hard_news.


### Detroit News local news

- Back to [Table of Contents](#Table-of-Contents)

Detroit News local news:

- `context_text/examples/articles/articles-TDN-local_news.py`
- local hard news sections (stored in `from context_text.collectors.newsbank.newspapers.DTNB import DTNB` - `DTNB.NEWS_SECTION_NAME_LIST`):

    - "Business"
    - "Metro"
    - "Nation" - because of auto industry stories

- in-house implementor (based on byline patterns, stored in `DTNB.Q_IN_HOUSE_AUTHOR`):

    - Byline ends in "/ The Detroit News", ignore case.

        - `Q( author_varchar__iregex = r'.*\s*/\s*the\s*detroit\s*news$' )`

    - Byline ends in "Special to The Detroit News", ignore case.

        - `Q( author_varchar__iregex = r'.*\s*/\s*special\s*to\s*the\s*detroit\s*news$' )`

    - Byline ends in "Detroit News * Bureau", ignore case.

        - `Q( author_varchar__iregex = r'.*\s*/\s*detroit\s*news\s*.*\s*bureau$' )`   

In [10]:
# filter queryset to just locally created Detroit News (TDN) articles.
# imports
from context_text.models import Article
from context_text.models import Newspaper
from context_text.shared.context_text_base import ContextTextBase
from context_text.collectors.newsbank.newspapers.DTNB import DTNB

# declare variables - Detroit News
do_apply_tag = False
tag_to_apply = None
tdn_local_news_sections = []
tdn_newspaper = None
tdn_article_qs = None
article_count = -1

# declare variables - filtering
include_opinion_columns = True
tags_in_list = []
tags_not_in_list = []
filter_out_prelim_tags = False
random_count = -1

# declare variables - make list of article IDs from QS.
article_id_list = []
article_counter = -1
current_article = None

# ==> configure

# configure - size of random sample we want
#random_count = 60

# configure - also, apply tag?
do_apply_tag = False
tag_to_apply = ContextTextBase.TAG_LOCAL_HARD_NEWS

# set up "local, regional and state news" sections
tdn_local_news_sections = DTNB.LOCAL_NEWS_SECTION_NAME_LIST

# Detroit News
# get newspaper instance for TDN.
tdn_newspaper = Newspaper.objects.get( id = DTNB.NEWSPAPER_ID )

# start with all articles
#tdn_article_qs = Article.objects.all()

# ==> filter to newspaper, local news section list, and in-house reporters.

# ----> manually

# now, need to find local news articles to test on.
#tdn_article_qs = tdn_article_qs.filter( newspaper = tdn_newspaper )

# only the locally implemented sections
#tdn_article_qs = tdn_article_qs.filter( section__in = tdn_local_news_sections )

# and, with an in-house author
#tdn_article_qs = tdn_article_qs.filter( DTNB.Q_IN_HOUSE_AUTHOR )

#print( "manual filter count: {}".format( tdn_article_qs.count() ) )

# ----> using Article.filter_articles()
tdn_article_qs = Article.filter_articles( qs_IN = tdn_article_qs,
                                          newspaper = tdn_newspaper,
                                          section_name_list = tdn_local_news_sections,
                                          custom_article_q = DTNB.Q_IN_HOUSE_AUTHOR )

print( "Article.filter_articles count: {}".format( tdn_article_qs.count() ) )

# and include opinion columns?
if ( include_opinion_columns == False ):
    
    # do not include columns
    tdn_article_qs = tdn_article_qs.exclude( author_string__in = DTNB.COLUMNIST_NAME_LIST )
    
#-- END check to see if we include columns. --#

'''
# filter to newspaper, section list, and in-house reporters.
tdn_article_qs = Article.filter_articles( qs_IN = tdn_article_qs,
                                          start_date = "2009-12-01",
                                          end_date = "2009-12-31",
                                          newspaper = tdn_newspaper,
                                          section_name_list = tdn_local_news_sections,
                                          custom_article_q = DTNB.Q_IN_HOUSE_AUTHOR )
'''

# how many is that?
article_count = tdn_article_qs.count()

print( "Article count before filtering on tags: " + str( article_count ) )

# ==> tags

# tags to exclude
#tags_not_in_list = [ "prelim_reliability", "prelim_network" ]
#tags_not_in_list = [ "minnesota1-20160328", "minnesota2-20160328", ]

# for later - exclude articles already coded.
#tags_not_in_list = [ OpenCalaisV2ArticleCoder.TAG_CODED_BY_ME ]

tags_not_in_list = None
if ( ( tags_not_in_list is not None ) and ( len( tags_not_in_list ) > 0 ) ):

    # exclude those in a list
    print( "filtering out articles with tags: " + str( tags_not_in_list ) )
    tdn_article_qs = tdn_article_qs.exclude( tags__name__in = tags_not_in_list )

#-- END check to see if we have a specific list of tags we want to exclude --#

# include only those with certain tags.
#tags_in_list = [ "prelim_unit_test_001", "prelim_unit_test_002", "prelim_unit_test_003", "prelim_unit_test_004", "prelim_unit_test_005", "prelim_unit_test_006", "prelim_unit_test_007" ]
#tags_in_list = [ "tdn_month", ]
tags_in_list = None
if ( ( tags_in_list is not None ) and ( len( tags_in_list ) > 0 ) ):

    # filter
    print( "filtering to just articles with tags: " + str( tags_in_list ) )
    tdn_article_qs = tdn_article_qs.filter( tags__name__in = tags_in_list )
    
#-- END check to see if we have a specific list of tags we want to include --#

# filter out "*prelim*" tags?
#filter_out_prelim_tags = True
if ( filter_out_prelim_tags == True ):

    # ifilter out all articles with any tag whose name contains "prelim".
    print( "filtering out articles with tags that contain \"prelim\"" )
    tdn_article_qs = tdn_article_qs.exclude( tags__name__icontains = "prelim" )
    
#-- END check to see if we filter out "prelim_*" tags --#

# how many is that?
article_count = tdn_article_qs.count()

print( "Article count after tag filtering: " + str( article_count ) )

# do we want a random sample?
if ( random_count > 0 ):

    # to get random, order them by "?", then use slicing to retrieve requested
    #     number.
    tdn_article_qs = tdn_article_qs.order_by( "?" )[ : random_count ]
    
#-- END check to see if we want random sample --#

# this is a nice algorithm, also:
# - http://www.titov.net/2005/09/21/do-not-use-order-by-rand-or-how-to-get-random-rows-from-table/

# make ID list, tag articles if configured to.
article_id_list = []
article_counter = 0
for current_article in tdn_article_qs:

    # increment article_counter
    article_counter += 1

    # add IDs to article_id_list
    article_id_list.append( str( current_article.id ) )
    
    # apply a tag while we are at it?
    if ( ( do_apply_tag == True ) and ( tag_to_apply is not None ) and ( tag_to_apply != "" ) ):
    
        # yes, please.  Add tag.
        current_article.tags.add( tag_to_apply )
        
    #-- END check to see if we apply tag. --#

    # output the tags.
    if ( debug_flag == True ):
        print( "- Tags for article " + str( current_article.id ) + " : " + str( current_article.tags.all() ) )
    #-- END DEBUG --#

#-- END loop over articles --#

# output the list.
print( "tdn_article_qs count: {}".format( tdn_article_qs.count() ) )
print( "Found " + str( article_counter ) + " articles ( " + str( article_count ) + " )." )
if ( debug_flag == True ):
    print( "List of " + str( len( article_id_list ) ) + " local TDN staff article IDs: " + ", ".join( article_id_list ) )
#-- END DEBUG --#


Article.filter_articles count: 13070
Article count before filtering on tags: 13070
Article count after tag filtering: 13070
tdn_article_qs count: 13070
Found 13070 articles ( 13070 ).


# Code Articles

- Back to [Table of Contents](#Table-of-Contente)

Retrieve just publications that are tagged as being local hard news and that also are not tagged as having been coded by OpenCalaisV2.

In [ ]:
# declare variables

# declare variables - article filter parameters
start_pub_date = None # should be datetime instance
end_pub_date = None # should be datetime instance
tags_in_list = []
tags_not_in_list = []
paper_id_in_list = []
section_list = []
article_id_in_list = []
params = {}

# declare variables - processing
do_i_print_updates = True
my_article_coding = None
article_qs = None
article_count = -1
coding_status = ""
limit_to = -1
do_coding = True

# declare variables - results
success_count = -1
success_list = None
got_errors = False
error_count = -1
error_dictionary = None
error_article_id = -1
error_status_list = None
error_status = ""
error_status_counter = -1

# first, get a list of articles to code.

# ! Set param values.

# ==> start and end dates
#start_pub_date = "2009-12-06"
#end_pub_date = "2009-12-12"

# ==> tagged articles

# Examples:
#tag_in_list = "prelim_reliability"
#tag_in_list = "prelim_network"
#tag_in_list = "prelim_unit_test_007"
#tag_in_list = [ "prelim_reliability", "prelim_network" ]
#tag_in_list = [ "prelim_reliability_test" ] # 60 articles - Grand Rapids only.
#tag_in_list = [ "prelim_reliability_combined" ] # 87 articles, Grand Rapids and Detroit.
#tag_in_list = [ "prelim_training_001" ]
#tag_in_list = [ "grp_month" ]

# ----> include articles when these tags are present.
#tags_in_list = None
tags_in_list = []
tags_in_list.append( ContextTextBase.TAG_LOCAL_HARD_NEWS )

# ---> exclude articles when these tags are present.
#tags_not_in_list = None
tags_not_in_list = []
tags_not_in_list.append( OpenCalaisV2ArticleCoder.TAG_CODED_BY_ME )

# ==> IDs of newspapers to include.
#paper_id_in_list = "1"

# ==> names of sections to include.
#section_list = "Lakeshore,Front Page,City and Region,Business"

# ==> just limit to specific articles by ID.
article_id_in_list = []
#article_id_in_list = [ 360962 ]
#article_id_in_list = [ 28598 ]
#article_id_in_list = [ 21653, 21756 ]
#article_id_in_list = [ 90948 ]
#article_id_in_list = [ 21627, 21609, 21579 ]
#article_id_in_list = [ 48778 ]
#article_id_in_list = [ 6065 ]
#article_id_in_list = [ 221858 ]
#article_id_in_list = [ 23804, 22630 ]
#article_id_in_list = [ 23804 ]

# filter parameters
params[ ArticleCoding.PARAM_START_DATE ] = start_pub_date
params[ ArticleCoding.PARAM_END_DATE ] = end_pub_date
params[ ArticleCoding.PARAM_TAGS_IN_LIST ] = tags_in_list
params[ ArticleCoding.PARAM_TAGS_NOT_IN_LIST ] = tags_not_in_list
params[ ArticleCoding.PARAM_PUBLICATION_LIST ] = paper_id_in_list
params[ ArticleCoding.PARAM_SECTION_LIST ] = section_list
params[ ArticleCoding.PARAM_ARTICLE_ID_LIST ] = article_id_in_list

# set coder you want to use.

# OpenCalais REST API v.2
params[ ArticleCoding.PARAM_CODER_TYPE ] = ArticleCoding.ARTICLE_CODING_IMPL_OPEN_CALAIS_API_V2

# get instance of ArticleCoding
my_article_coding = ArticleCoding()
my_article_coding.do_print_updates = do_i_print_updates

# set params
my_article_coding.store_parameters( params )

print( "Query Parameters: {}".format( params ) )

# create query set - ArticleCoding does the filtering for you.
article_qs = my_article_coding.create_article_query_set()

print( "After my_article_coding.create_article_query_set(), count: {}".format( article_qs.count() ) )
if ( article_qs._result_cache is None ):
    
    print( "article_qs evaluated: NO ( {} )".format( article_qs._result_cache ) )
    
else:
    
    print( "article_qs evaluated: YES" )

#-- END check to see if _result_cache --#

# limit for an initial test?
limit_to = 4990
if ( ( limit_to is not None ) and ( isinstance( limit_to, int ) == True ) and ( limit_to > 0 ) ):

    # yes.
    article_qs = article_qs[ : limit_to ]

#-- END check to see if limit --#

# get article count
if ( isinstance( article_qs, list ) == True ):

    # list - call len()
    article_list = article_qs
    article_count = len( article_list )
    
else:

    # not a list - call count()
    article_count = article_qs.count()
    
#-- END figure out how to get count --#

print( "Matching article count: " + str( article_count ) )

# Do coding?
if ( do_coding == True ):

    print( "do_coding == True - it's on!" )

    # yes - make sure we have at least one article:
    if ( article_count > 0 ):

        # invoke the code_article_data( self, query_set_IN ) method.
        coding_status = my_article_coding.code_article_data( article_qs )
    
        # output status
        print( "\n\n==============================\n\nCoding status: \"" + coding_status + "\"" )
        
        # get success count
        success_count = my_article_coding.get_success_count()
        print( "\n\n====> Count of articles successfully processed: " + str( success_count ) )    
        
        # if successes, list out IDs.
        if ( success_count > 0 ):
        
            # there were successes.
            success_list = my_article_coding.get_success_list()
            print( "- list of successfully processed articles: " + str( success_list ) )
        
        #-- END check to see if successes. --#
        
        # got errors?
        got_errors = my_article_coding.has_errors()
        if ( got_errors == True ):
        
            # get error dictionary
            error_dictionary = my_article_coding.get_error_dictionary()
            
            # get error count
            error_count = len( error_dictionary )
            print( "\n\n====> Count of articles with errors: " + str( error_count ) )
            
            # loop...
            for error_article_id, error_status_list in six.iteritems( error_dictionary ):
            
                # output errors for this article.
                print( "- errors for article ID " + str( error_article_id ) + ":" )
                
                # loop over status messages.
                error_status_counter = 0
                for error_status in error_status_list:
                
                    # increment status
                    error_status_counter += 1

                    # print status
                    print( "----> status #" + str( error_status_counter ) + ": " + error_status )
                    
                #-- END loop over status messages. --#
            
            #-- END loop over articles. --#
   
        #-- END check to see if errors --#
    
    #-- END check to see if article count. --#
    
else:
    
    # output matching article count.
    print( "do_coding == False, so dry run" )
    
#-- END check to see if we do_coding --#


Query Parameters: {'start_date': None, 'end_date': None, 'tags_in_list_IN': ['local_hard_news'], 'tags_not_in_list_IN': ['coded-OpenCalaisV2ArticleCoder'], 'publications': [], 'section_list': [], 'article_id_list': [], 'coder_type': 'open_calais_api_v2'}
After my_article_coding.create_article_query_set(), count: 43254
article_qs evaluated: YES
Matching article count: 4990
do_coding == True - it's on!


==> article 1: 377049 - Frontline Community Church


==> article 2: 377011 - Soldier turned pastor uses modern weapons in - A war for souls


==> article 3: 377069 - Series explores religion, racial misconceptions - Racial assumptions can skew Scripture, a Hope professor says


==> article 4: 377088 - "The Fruitful Life"


==> article 5: 376913 - Church unites behind challenge of evangelism - "The Fruitful Life," a 40-day study of what it means to bear fruit as believers
in Christ



==> article 6: 376924 - Authors offer soaring tales of rare birds


==> article 7: 376946 - Relief worker



==> article 50: 377851 - Hope a long shot to win the MIAA title - Flying Dutchmen will improve, but not enough


==> article 51: 377672 - Hope trio coming into its own - Boles, Wood, Henderson help team win tourney


==> article 52: 377841 - Rough road - Entering MIAA play, hope trying to find way out of funk


==> article 53: 377954 - EGR native links Australian, U.S. firms - New trade agreement will benefit both countries, Christine Gibbs-Stewart says


==> article 54: 378102 - Deal with Swedish firm helps Inrad - Kentwood company expects to add jobs


==> article 55: 378202 - Riviera's report raises warning - Auditor doubts company can continue after bigger-than-expected loss


==> article 56: 378092 - Bargain hunting


==> article 57: 378007 - Estate sale lures collectors


==> article 58: 378049 - 'He just wanted to have a good time' - Police say Holland man was slain after an argument that started in a Grand Rapids bar


==> article 59: 378006 - Resolutions you can handle - For



==> article 99: 378293 - District to raze ice cream shop - Frosty Boy lot will become storage, parking for Creston, Palmer


==> article 100: 378416 - Mining project turned down


==> article 101: 378462 - Township sued over zoning - Restoration of residential status sought


==> article 102: 378251 - Howard Stern yanked - Shock jock dropped from local station, apparently over tendency to promote his future employer


==> article 103: 378375 - Focus shifts to finding killer - State police say 23-year-old Julia Dawson was murdered


==> article 104: 378430 - Howard Stern yanked - Shock jock dropped from local station, apparently over tendency to promote his future employer


==> article 105: 378504 - Focus shifts to finding killer - State police say 23-year-old Julia Dawson was murdered


==> article 106: 378536 - Tragedy can result if kids don't learn danger of fire - Experts say parents tend to ignore youngsters' fascination with flames


==> article 107: 378496 - Playing with fire




==> article 147: 379249 - Mentor brings FIRST to alternative high - The Nonconformants are registered to compete in two tournaments


==> article 148: 379223 - Competitive cheer


==> article 149: 378911 - Past stars shine again for Hope - Top women alumni face JV basketball team Saturday


==> article 150: 378809 - Panthers cheer team seeks return to top - Squad finished fifth at state after making top three previous four years


==> article 151: 378756 - Flying Dutchmen's fresh start - Hope men open MIAA play with convincing victory over Tri-State


==> article 152: 379023 - Family bowling tournament puts priority on fun - Event pairing child with adult bowlers will held during next two weekends


==> article 153: 378652 - It's time to bury 2004 catch phrases


==> article 154: 379374 - Shoppers gained steam - Gift cards were season's biggest hit


==> article 155: 379475 - Businesswoman buys into neighborhood revival - Antiques store becomes shopping destination


==> article 156:



==> article 195: 379903 - Learning to give, learning to accept


==> article 196: 379712 - Ministry uses culture to share faith


==> article 197: 379852 - Without answers, ask 'how' not 'why' - There often is meaning in response to tragedy


==> article 198: 379885 - Local faith groups join in relief effort


==> article 199: 379651 - Through teamwork, tragedies are overcome


==> article 200: 379793 - Editorial helps to get CRC readers' goats


==> article 201: 380062 - Our voices heard -- even in Spanish - State commission now includes four from area


==> article 202: 380433 - A step above - Wolverine set pace in good year for local stocks


==> article 203: 380378 - Auto Show's theme: anything (that) goes - GM unveiling three advanced hybrid vehicles


==> article 204: 380018 - 'Hot Dog Man' feeds need for GR's hungry - It's a humble menu. But, for those who line up early each week, it's a lifeline


==> article 205: 380071 - Lunch money


==> article 206: 380147 - School will t



==> article 236: 380913 - Family gathering - Werley siblings host benefit for sister, but she dies days before event
ValueError parsing OpenCalais JSON for Article 380913 - raw response body: You exceeded the concurrent request limit for your license key. Please try again later or contact support to upgrade your license.
======> In code_article_data(): ERROR - Error: ValueError parsing OpenCalais JSON for Article 380913 - raw response body: You exceeded the concurrent request limit for your license key. Please try again later or contact support to upgrade your license.; article = 380913 - Jan 10, 2005, Lakeshore ( L3 ), UID: 10796389890F8644 - Family gathering - Werley siblings host benefit for sister, but she dies days before event ( Grand Rapids Press, The )


==> article 237: 380664 - 'Rusty' Grand Haven cruises to wrestling title - Ten Bucs wrestlers reach finals of invitational


==> article 238: 381002 - Pen Club OKs sale of health club - $1 million deal puts downtown athletic 



==> article 276: 381436 - "I'll Be There"


==> article 277: 381607 - Voters asked to restore millage


==> article 278: 381355 - Activist says gays steal civil rights rhetoric - Rev. Eugene Rivers believes campaign for gay marriage is assertion of white
privilege



==> article 279: 381307 - High school student's song honored by music teachers


==> article 280: 381313 - Plant to be demolished after multiple delays - Demolition deal scrapped, could lead to lawsuit


==> article 281: 381406 - City may raise developers' fees - Increase revenue could pay for two inspectors


==> article 282: 381316 - Meijer will donate land to Spectrum - Site of first Meijer store could become community center


==> article 283: 381369 - 22 years of tax breaks proposed for condo project - Developer would remodel West Middle School


==> article 284: 381324 - Learning curve: County ponders roundabouts - Statistics show circular intersections prevent crashes


==> article 285: 381508 - On the street, rea



==> article 323: 382290 - New charter high school studied - National heritage looks for way to meet demand


==> article 324: 381822 - Teen says boyfriend not violent - Wayland girl says his e-mails, allegedly threatening school massacre, were misunderstood


==> article 325: 382118 - School districts don't make grade - Grand Rapids, Kentwood, Wyoming, 6 more miss federal standards


==> article 326: 382211 - Rain, rain, it's going away; winter weather returns today - Warm air, fog, lightning give way to cold and snow


==> article 327: 381981 - Funds sought for 'fun' spin on bridge work - City lobbied to give $50,000 for campaign about project


==> article 328: 382050 - Hospital launches second annual fitness challenge - Schools take part in wellness program


==> article 329: 382301 - Township clerk resigns, heads South - Bob Lamar's pick as his replacement is OK'd by board


==> article 330: 382084 - Kickoff


==> article 331: 382074 - Students take whack at colonial life - Stude



==> article 371: 382717 - Pass over state judges - Former governor says senators blocking nominees


==> article 372: 382518 - Engler's says he'd rather skip Michigan nominees than endure more judicial gridlock


==> article 373: 382883 - Local cafe-juice bar aims to sustain - D. Marie's Cafe uses local suppliers to support area economy


==> article 374: 382771 - Duba's to leave East Beltline by fall - Restaurant looks for new building as bank, hotel built on former site


==> article 375: 383063 - Economy is improving -- slowly - Local economist discusses outlook for next decade


==> article 376: 383081 - Feds seek plea deal in fraud case - Government says advocate for disabled operated ponzi scheme


==> article 377: 382998 - Unite to make dream reality, leaders say


==> article 378: 382757 - 3 vie for 5th Ward council seat - Holland City Council will chose Wednesday


==> article 379: 383005 - State of City speech echoes King


==> article 380: 382795 - Historic long johns may 



==> article 419: 383525 - DePree found his niche at Albion - Holland grad a possible MVP in the MIAA


==> article 420: 383198 - Milobinski has West Ottawa swimming on track - Panthers won big meet at Rockford last week


==> article 421: 383659 - Travis DePree file


==> article 422: 383392 - Job future hinges on Big Three - As Gov. Granholm put it, state remains 'global center' for auto production and research


==> article 423: 383770 - 29th St. fire draws crowd - Smoke billowing from blaze at car dealership building seen for miles


==> article 424: 383753 - Surgeons use band as weight-loss alternative - Procedure is less invasive than gastric bypass


==> article 425: 383826 - Academic boosters pay teacher-tutors - High school students study with educator to get ready for semester exam week


==> article 426: 383849 - Boosters pay teacher-tutors - Educators help students get ready for exam week


==> article 427: 383902 - Lap Band procedure


==> article 428: 383968 - Campus cha



==> article 468: 384107 - For sale: Two school buildings - School district seeks buyer for properties near downtown


==> article 469: 384089 - No landslide for vote machine - Some election officials dispute choice for countywide system


==> article 470: 384394 - Book honors delight Calvin professor - Gary Schmidt's 'Lizzie Bright and the Buckminister Boy' wins Newbery Honor


==> article 471: 384301 - Marching for equality - Dozens brave cold, snow to honor civil rights leader


==> article 472: 384303 - City will expand cemetery - Council rejects plan to develop second graveyard


==> article 473: 384315 - Out-of-town firm narrowly wins maintenance contract - Council's 4-3 vote follows decision not to favor local bidders


==> article 474: 384349 - Officials interview 5th Ward candidates - Council will make its selection Wednesday


==> article 475: 384080 - Longtime board member to step down


==> article 476: 384169 - Township's first supervisor, Gerald Michmerhuizen, dies


==>



==> article 514: 384545 - New estimate may delay road projects


==> article 515: 384575 - Deputy gives township update on crime fighting


==> article 516: 384751 - Day care's proposed site worries officials - Traffic concerns planning commissioners


==> article 517: 384761 - Blackhawks coach earns honor - He is regional football coach of year


==> article 518: 384471 - Hope loads up with defense - Flying Dutchmen basketball team's lineup change has top scorers coming off bench


==> article 519: 384695 - Dutch wrestlers edge Panthers - Holland held advantage in heavier weight classes


==> article 520: 385233 - Mid-winter dreams - Auto show reveals future cars, celebrates past


==> article 521: 385120 - Kentucky fire stalls Electrolux production - More than 1,000 employees will be sidelined up to six weeks


==> article 522: 384829 - Jobless rate climbs - Rise 'not what we were expecting,' economist says


==> article 523: 384972 - Seventh annual Michigan International Auto Show



==> article 563: 385795 - River iced over for almost 18 miles - Roller coaster weather blamed for flood-causing ice jams


==> article 564: 385798 - 'Amazing things' ahead, New zoo director says - Neighbor's group leader reacts favorably to choice of Shedd Aquarium exec


==> article 565: 385780 - Witness tells how argument led to killing - Marvis Roberts is ordered to stand trial in slaying outside Northeast apartment


==> article 566: 385629 - The Bert Vescolani file


==> article 567: 385649 - Approval for dispatch changes delayed - City Commission worries about safety if civilians answer fire calls


==> article 568: 385712 - Learn Arabic


==> article 569: 385533 - Lawyer: Utility 'irresponsible' about Ada line - Iserv CEO suing Consumers Energy over its refusal to bury line


==> article 570: 385748 - Arabic program for young students begins - Classes taught by former teacher, interpreter


==> article 571: 385574 - Library upgrades book scanning - New book tags save time by r



==> article 609: 386728 - China puts its stamp on local economy - $50-a-month labor, protective government bring shift in our outlook


==> article 610: 386820 - China question - Do you go or make your stand in West Michigan?


==> article 611: 386436 - Job growth mostly lower paying - Fewer companies expect to hire management


==> article 612: 386538 - Engler pushes agenda for manufacturers - Former Michigan governor looks to raise lobby's political profile


==> article 613: 386434 - Winter takes GR by storm - Heavy snowfall makes for slick roads, cancellations


==> article 614: 386687 - What history hides in your walls? - Woman shares documents dating back 145 years


==> article 615: 386504 - Healthy Van Houten sextuplets turn 1 - Their parents say tots have passed important milestones


==> article 616: 386345 - The Van Houten Sextuplets


==> article 617: 386701 - Holland's Accents shuts its doors - It has operated in Tower Clock since 1992


==> article 618: 386813 - ROTC co



==> article 647: 387399 - Couple hit the road with asphalt business - New equipment will allow firm to use recycled asphalt to repair potholes
ValueError parsing OpenCalais JSON for Article 387399 - raw response body: You exceeded the concurrent request limit for your license key. Please try again later or contact support to upgrade your license.
======> In code_article_data(): ERROR - Error: ValueError parsing OpenCalais JSON for Article 387399 - raw response body: You exceeded the concurrent request limit for your license key. Please try again later or contact support to upgrade your license.; article = 387399 - Jan 25, 2005, Business ( C1 ), UID: 107E51BE3129B21C - Couple hit the road with asphalt business - New equipment will allow firm to use recycled asphalt to repair potholes ( Grand Rapids Press, The )


==> article 648: 387425 - Insurer to pay for mistake - Car owners will get cash for overpaying for salvaged vehicles


==> article 649: 387148 - Knape and Vogt sales up, but 



==> article 686: 387870 - Utility memo gives look at plan for power line - Internal document presented in court shows Consumers Energy expected to battle
residents



==> article 687: 387806 - Robber not playing, victim learns - Party store worker wounded by what he thought was toy gun


==> article 688: 387861 - Utility memo gives look at plan for power line - Internal document presented in court shows Consumers Energy expected to battle
residents



==> article 689: 387761 - No bail allowed in bomb threat - Cedar Springs man will wait in D.C. jail while grand jury considers case


==> article 690: 387694 - No bail allowed in bomb threat - Cedar Springs man will wait in D.C. jail while grand jury considers case


==> article 691: 387555 - Bleke has idea to honor King - Board asked to consider naming rebuilt school for slain civil rights leader


==> article 692: 387652 - Follow the leader, but which one?


==> article 693: 387794 - What's in a name?


==> article 694: 387708 - Older



==> article 733: 388137 - Misguided alert causes tunnel trouble - Two lanes will reopen after frozen pipe is fixed


==> article 734: 388013 - City officials intrigued by proposals for Holland station - A condo project would remove the landmark


==> article 735: 388343 - Boys and Girls Club adds north-side program


==> article 736: 388559 - Hawkeyes' new tricks help cheer team rack up points - Stunts have been key improvement for squad


==> article 737: 388455 - Quick starter - Vilmann sparks big lead, easy win for flying dutch


==> article 738: 388001 - Defenseman provides offense in Panthers' win - Hockey team holds off late rally by Hudsonville


==> article 739: 388103 - California dreamin' on a Michigan winter's day


==> article 740: 388433 - Web site backs Watkins, pans Granholm - State superintendent says he will ask it to be taken down


==> article 741: 388674 - Love affair with cars goes on - Mustangs, vans, stretch limo Volkswagen Beetles wow visitors at show


==> ar



==> article 783: 389008 - Lawyer says police 'promises' led to lie - Witness faces life in prison for perjury in Deanna Gillean case


==> article 784: 389240 - Hope exceeds fund-raising target - More than 3,300 donors contributed to help with such projects as DeVos Fieldhouse


==> article 785: 389119 - Governor says economy needs more than tax cuts - She advocates other improvements while promoting business tax plan


==> article 786: 389232 - Kenowa school board to review contract offer Monday - Union hopeful agreement can be reached


==> article 787: 389077 - Trial ordered for man accused of stabbing - He also was a 'person of interest' in GR cop shooting case


==> article 788: 389180 - Man shot by police pleads no contest to assault - Deputy fired after suspect allegedly tried to run him over with truck


==> article 789: 389238 - Forest Hills cuts 34 part-time jobs - Substitute teachers, custodians and support staff lose jobs


==> article 790: 389123 - Holland leaders retrea



==> article 830: 390066 - New tower on horizon - Alticor shows off plans for downtown Marriott, which would be city's third-tallest building


==> article 831: 390031 - New tower on horizon - Alticor shows off plans for downtown Marriott, which would be city's third-tallest building


==> article 832: 390168 - Changing the skyline


==> article 833: 390029 - City keeps dreams realistic - Council members remember economic woes when setting year's goals


==> article 834: 390011 - Hope seminars combine learning and fun - Winter Happening brings college closer to community


==> article 835: 390023 - School speed signs may be enlarged - Road commissioners worry drivers can't see speed limits


==> article 836: 390209 - Chicken rice soup, hot bread make for good winter meal


==> article 837: 390056 - Black River eyes league title - River Rats basketball team leads Alliance


==> article 838: 389976 - Snow makes running tough in biathlon - Former Hope students win Pigeon Creek event


==



==> article 874: 390699 - Report: GR police aren't brutal - Officers used force in 551 incidents last year -- less than 4 percent of its 14,000 arrests


==> article 875: 390618 - Persistence pays off in sand-mining fight - Homeowners sue township, win reduction in amount to be removed


==> article 876: 390551 - DeVos Place ballroom opens with fanfare - Workers applaud those who got ball rolling for convention center


==> article 877: 390767 - Full stomachs - Area schools feed students to prevent testing distractions


==> article 878: 390882 - Minorities needed for blood, marrow donation - Drive held to kick off Black History Month


==> article 879: 390720 - FHN grad's ship tossed on Pacific - Jenna Klotz and Semester at Sea group now are safe


==> article 880: 390680 - West misunderstands Russia, Ike's granddaughter says
ValueError parsing OpenCalais JSON for Article 390680 - raw response body: You exceeded the concurrent request limit for your license key. Please try again lat



==> article 913: 391573 - Judge delivers jolt in power case - He says injunction would harm electricity users in growing suburbs


==> article 914: 391021 - Festival features chili, sculptures - Ice-carving contest is highlight of 3-day event


==> article 915: 391520 - GRCC celebrates its future - Anniversary marked by new mural, pilot's speech


==> article 916: 391598 - Council moves ahead on Baker factory lofts - Project still has risk, developer admits


==> article 917: 391495 - Sikkema: Schools can ask for tax hike - Senate majority leader says local option exists for increasing education funding


==> article 918: 391562 - Wyoming man blows whistle on big game


==> article 919: 391203 - Forum hears residents share accusations of abuse by police - Grand Rapids' Community Relations Commission hears complaints
about police methods



==> article 920: 391530 - Ice Breaker features chili, sculptures - Ice-carving competition is highlight of three-day event


==> article 921: 3914



==> article 959: 391821 - Friends seek peace pole to remember activist - Jon Christian Den Herder died of cancer Wednesday


==> article 960: 391915 - Cash-strapped department to skip fire investigation - City has delayed filling fire marshal position


==> article 961: 392025 - Garcia calls for school funding change - Superintendent tells Rotary members to pressure lawmakers for solution


==> article 962: 392006 - Hope's Ebels overcomes injury, foul trouble - Coach keeps post player fresh for stretch run


==> article 963: 392012 - Depth leads Bucs to win - Grand Haven's underclassmen help beat Maroons in swimming dual meet


==> article 964: 391853 - Hawkeyes handle Broncos despite weak effort - Coach says volleyball squad didn't play with heart


==> article 965: 391863 - Bucs star to play soccer at Cornell - Forward will compete in Division I Ivy League Big Red


==> article 966: 391847 - Overseas volunteers should start at home - Churches need to probe their reasons


==> artic



==> article 1006: 392879 - School where boy fell from rope gets gym climbing wall as gift - PTA raised more than $9,000 after Sam Barwin was hurt


==> article 1007: 392508 - Paganellis make officiating a family affair - Relatives gather to watch game -- and their guy in zebra stripes


==> article 1008: 392453 - Converging worlds - Interracial adoption becomes more prevalent due to shortage of black families


==> article 1009: 392936 - Living in exile - Woman will leave America to join fiance she believes was wrongly deported


==> article 1010: 392971 - Transracial Adoption Fair


==> article 1011: 392784 - Official duty


==> article 1012: 392647 - District extends background checks to student teachers - Demand for police, FBI report went into effect this semester


==> article 1013: 392600 - Lunch and money - Director uses student ideas to shore up program


==> article 1014: 392502 - Dean of law school to speak at summit - Frank Wu is a leading voice in the Asian-American commu



==> article 1052: 393416 - Old-barn lovers see cathedrals - Passions enflamed, preservationist willing to try to find home for structure rather than
see it burn



==> article 1053: 393594 - Donut heaven


==> article 1054: 393573 - A paczki by any name delicious - Call them what you will, huge filled donuts herald arrival of Fat Tuesday


==> article 1055: 393446 - Situation critical for missing diabetic - Joseph Gillis was last seen at two locations on Friday


==> article 1056: 393321 - Teacher keeps job, changes schools - Controversial book will not be banned


==> article 1057: 393370 - GRCC tuition expected to rise again - Out-of-district students bear the brunt of jump


==> article 1058: 393378 - College housing disparity - Dorms dingy, typical president home worth $933,000


==> article 1059: 393422 - Russia still growing toward democracy, expert says - Susan Eisenhower calls accusations against Putin's regime 'exaggerated'


==> article 1060: 393528 - Russia still growing t



==> article 1094: 393934 - Town hall meetings


==> article 1095: 393648 - Sides square off over Jade Pig project - Referendum Feb. 22 on high-rise condos


==> article 1096: 393742 - Commissioners: Budget survey flawed


==> article 1097: 393736 - Knapp's Corner car wash denied - Commissioner doesn't want 'another Alpine'


==> article 1098: 393642 - Former Hope leader dies - John H. Jacobson, who served as president from 1987-99, died Tuesday


==> article 1099: 393814 - Firefighters accused of abusing policies - GR fire chief, city manager say last-minute vacations, sick leave bleed budget
reserves dry



==> article 1100: 393911 - First 'State of County' addresses challenges - Ottawa faces budget troubles, but administrator lauded for stellar first year


==> article 1101: 393785 - Doctor pioneered colorectal surgery - 'Dr. Jim,' who founded Ferguson Clinic, died Sunday in Georgia


==> article 1102: 393867 - Jacobson helped Hope grow - Former president, who died Tuesday, guided 



==> article 1134: 394213 - Is countywide bus on its way? - Survey will determine service needs


==> article 1135: 394023 - GVSU faculty to gain royalties for work - All commercial research earnings had gone to university


==> article 1136: 394035 - Choir invited to sing anthem at Pistons game


==> article 1137: 394252 - Oh, say, can you sing?


==> article 1138: 394414 - Students' housing bills rise - Costs even higher for private rooms


==> article 1139: 394551 - Pine Rest adds space for kids in the pits - New hospital wing encourages socializing


==> article 1140: 394055 - Chilly night in Lansing - Observers wonder if more than politics divides Granholm, Sikkema


==> article 1141: 394559 - Granholm asks more funding for local schools - High schools would get bigger boost; colleges face cuts


==> article 1142: 394510 - Granholm asks more funding for local schools - High schools would get bigger boost; colleges face cuts


==> article 1143: 394590 - Heroin dealer says he and c



==> article 1180: 395138 - Rare complication kills activist - GRTV's Koning, recalled as 'a giant in every way'


==> article 1181: 394932 - Rescue services are city's top priority - Officials may consider consolidating ambulance and fire departments


==> article 1182: 394979 - Students seek comedic payoff - Actors tickle funny bones for shot at state title


==> article 1183: 395124 - Developer punished for cutting down restricted trees - Kelly Ridge Estates fined, new trees to be planted


==> article 1184: 394959 - County extends fairground lease - Move opens door for expansion


==> article 1185: 394984 - Flying Dutch coach may need antacids for Saturday - Game with Tri-State is expected to be high-scoring affair


==> article 1186: 394817 - Dominating day one - Hope's men's swim team takes commanding lead in MIAA championships


==> article 1187: 395014 - Grand Haven wrestlers sweep dual season - Team easily handles Panthers before Saturday's O-K Red Lakeshore meet


==> articl



==> article 1225: 395727 - 'Alegria Latina' on your AM dial - Longtime host excited about expansion, future of show for Hispanics


==> article 1226: 395905 - City agrees to add plots at cemetery - As many as 244 burial sites will be added


==> article 1227: 395868 - Multipurpose pavilion planned for park - It would replace former Donkersloot building


==> article 1228: 395719 - Holland duo sets sail in search of spot on Olympic team - Thinschmidt, Wierda look to qualify in 'tornado'


==> article 1229: 395402 - On the mend - Hope's Phillips coming along strong from finger injury


==> article 1230: 395430 - Zeeland bowler tops in league - Smith carries 187 average for Chix squad


==> article 1231: 395598 - Freshman making waves for Bucs - Heyblom aims to qualify for nationals in 100


==> article 1232: 395701 - No time to play politics - Granholm, Republicans must work together to keep state a leader


==> article 1233: 396126 - Meijers give to hospital's growth - New wing in Gre



==> article 1270: 396452 - Bar scrape means jail for suspect in hit-run - Man accused of killing girl in parking lot has bond revoked after he is charged
with assault



==> article 1271: 396252 - Bar scrape means jail for suspect in hit-run - Man accused of killing girl in parking lot has bond revoked after he is charged
with assault



==> article 1272: 396494 - Years of sacrifice pay off: 'Family has a home again' - Son's basketball team shows support as Bosnian family becomes U.S.
citizens



==> article 1273: 396340 - Survey results please officials - 95 percent said township was 'excellent' or 'good' place to live


==> article 1274: 396402 - BPW considers its options - Officials say coal-powered generating unit at downtown plant would meet needs


==> article 1275: 396284 - Learn about online predators - Police distribute free copies of CD-ROM for parents at city's three high schools


==> article 1276: 396317 - Millage to keep library privileges on May ballot - Saugatuck-Doug



==> article 1310: 396975 - The Baker Lofts


==> article 1311: 396732 - Housing project hits roadblock again - Commissioners worried about homes on proposed private road


==> article 1312: 396809 - Wisconsin Lutheran to leave MIAA - League still will get automatic berth in football playoffs


==> article 1313: 396899 - Panthers pitch-in - 12 Score for West Ottawa in easy win over Dutch


==> article 1314: 396749 - Flying Dutch coach recounts highs, lows of first 199 wins - He goes for win No. 200 tonight at Rochester


==> article 1315: 396866 - Chix outlast Rockets in double overtime - Steve Reidsma pushes game to OT, then sinks winner


==> article 1316: 396915 - Loss leaves Flying Dutchmen in spoiler role - Hope plays rival Calvin at Civic Center tonight


==> article 1317: 397232 - History won't stop new hotel - Preservation panel says downtown Israels building not historically significant


==> article 1318: 397190 - NHL meltdown leaves sports bars high and dry - Cancellation o



==> article 1350: 397306 - Animal-control worker fired


==> article 1351: 397488 - School vision, hearing checks may be cut - Health officials say eliminating screenings could hurt kids


==> article 1352: 397915 - Recycled shelving spurt pays off for company - Clark Associates MH grows retail customer base


==> article 1353: 398048 - Home guru says harmony, creativity equal happiness - Best-selling author/architect tells crowd to keep in mind 'what's good
for the planet'



==> article 1354: 397982 - So far, colleges spared from cuts - Senate rejects governor's plan, saying spending levels were guaranteed by law last year


==> article 1355: 397830 - Love of teaching lets professor push phobia aside - Jim Jadrich may be terrified of people, but his devotion to students prevails


==> article 1356: 398003 - Police forced to play tag with city's graffiti artists - Cops conduct tour of the works they call vandalism


==> article 1357: 398113 - Colleges spared from cuts so far - Senat



==> article 1397: 398241 - An eye on the future


==> article 1398: 398276 - Statute tied hands in priest case - Authorities unable to prosecute alleged sex abuse because clock ran out


==> article 1399: 398289 - Bernard trial still on, documents out - Judge refuses to remove prosecutors from case


==> article 1400: 398264 - What happened to Grand River rock? - Speculation ranges from it being swept away or sheared off


==> article 1401: 398375 - South Division Avenue may get commemorative designation - Jack Magnuson Way, Division Parkway are two names suggested


==> article 1402: 398162 - Church will remove old barn - It's too expensive for church to repair


==> article 1403: 398266 - West Ottawa officials mull visit from controversial author - Chris Crutcher's book was banned from Grand Rapids schools


==> article 1404: 398341 - Police crack down on Goodwill stealing - Stings have netted arrests of people taking off with donated items


==> article 1405: 398204 - Going bankru



==> article 1444: 399040 - 3 W. Michigan schools succeed in theater meet - West Ottawa, Holland and East Grand Rapids high schools among best


==> article 1445: 399181 - Teacher contract debates will rage again soon - Contracts expire in 12 area districts in June


==> article 1446: 399241 - Teacher contract debates will rage again soon - Contracts expire in 12 area districts in June


==> article 1447: 399057 - Novel's inspiration no mystery


==> article 1448: 399101 - Teacher contract debates to resume - Contracts expire in 12 area districts in June


==> article 1449: 399051 - Ice is nice - Festival draws crowds to shop, hunt for diamonds


==> article 1450: 399140 - Scouts take visitors around the world - World Thinking Day, which encourages girls to learn about other countries, started
75 years ago



==> article 1451: 399298 - Top 10 priorities


==> article 1452: 399281 - County addresses parks master plan - While continuing land acquisition efforts, emphasis is on upgrading



==> article 1490: 399567 - Staying current helps coach keep winning - Patience, knowledge of game add up to success for Maroons' Mike Phelps


==> article 1491: 399357 - Coach sees regional title for wrestlers - Bucs must overcome Hudsonville to advance after seven district championships


==> article 1492: 399306 - Gentex gets city's approval for industrial development district - Company wants to expand tax break to cover new facilities,
existing plant



==> article 1493: 399444 - CMH may cut funds to outreach program - Money would be used to help Girls on the Run


==> article 1494: 399649 - Office, retail development planned for Comstock Street


==> article 1495: 399717 - Drowning in debt - Medical bills, credit card debt fuel record bankuptcy filings


==> article 1496: 399765 - Toe-tapping device may help amputees - Company is developing sensor-loaded insole to send signals to prosthetic hand


==> article 1497: 399909 - Voters: Let condos rise high - Jade Pig's development wi



==> article 1539: 400725 - Museum a possible target of city budget ax - The venue's $1.9 million subsidy could be the target of cost savings


==> article 1540: 400411 - Christian schools offer discount - Financial benefits for younger students aim to stop enrollment decline


==> article 1541: 400502 - Jurisdictional arrest rules clarified - Charges dismissed against man arrested on bench warrant outside city


==> article 1542: 400158 - Airport settles rent dispute with FAA


==> article 1543: 400598 - Democratic chairwoman runs for NEA slot - She says campaign travel is challenging


==> article 1544: 400761 - Booming ridership leads GVSU to double Saturday bus runs - Free service is convenient way for students to experience city
life



==> article 1545: 400239 - Misdemeanor plea is approved in Saugatuck jet-boat death - Victim's family doesn't want son's friend sent to prison


==> article 1546: 400278 - Traffic in the front yard - Reconstruction of intersection threatens to bri



==> article 1579: 400789 - Kissinger comes


==> article 1580: 400955 - Kissinger to speak on China - Organizers say he is coming as favor to Ford


==> article 1581: 400824 - Foundation awards local grants


==> article 1582: 401011 - CRC official rests after cancerous tumor removed - Doctors are unsure when the Rev. David Engelhard will return to work


==> article 1583: 400775 - Why won't these performers play downtown? - Effort launched to get more minority acts to arena and DeVos Hall


==> article 1584: 400981 - Disabilities may have cost officer his job - Lawsuit claims ADD led to mistakes on time sheet


==> article 1585: 400991 - Sending prayers of hope - West Michigan Catholics offer wishes for pope's health
ValueError parsing OpenCalais JSON for Article 400991 - raw response body: You exceeded the concurrent request limit for your license key. Please try again later or contact support to upgrade your license.
======> In code_article_data(): ERROR - Error: ValueError parsin



==> article 1620: 401604 - For some reason, few minorities camp - State tries to change traditions but has little success


==> article 1621: 401606 - NAACP leader: 'Education can be a great avenue' - Political leaders encouraged youth at black history program


==> article 1622: 401576 - NAACP leader: 'Education can be a great avenue' - Political leaders encouraged youth at black history program


==> article 1623: 401905 - 'Outhouse 500' offers loads of fun - Coopersville Chill Celebration featured a parade, chili contest and races on Saturday


==> article 1624: 401757 - No-shows don't rattle officers over drug tests - Despite poor attendance, they aim to educate parents about prevention


==> article 1625: 401856 - Pluto rightfully deserves its planetary honors - Search for solar system's 9th planet began in 1905, took 25 years


==> article 1626: 401986 - Baby unknowingly shared life


==> article 1627: 401849 - Staking their claims


==> article 1628: 401999 - Baby unknowingly 



==> article 1666: 402194 - Versatile muffin mix handy to have around the house


==> article 1667: 402227 - GOP targets teacher benefit costs - Soaring health-care and retirement expenses raise red ink flags


==> article 1668: 402561 - Going back to school - Steelcase, American Seating team up to produce desk for colleges


==> article 1669: 402719 - Riviera stock soars - It's quite a turnaround for company an auditor said might not survive


==> article 1670: 402552 - Old building gets new life - Water plant becomes Clearwater Place


==> article 1671: 402502 - Stir-fry restaurant to open in Centerpointe Mall - Dunham's Sports also joins mall lineup


==> article 1672: 402454 - Stir-fry restaurant to open in Centerpointe Mall - Dunham's Sports also joins mall lineup


==> article 1673: 402431 - Going back to school - Steelcase, American Seating team up to produce desk for colleges


==> article 1674: 402410 - Old building gets new life - Water plant becomes Clearwater Place


==> a



==> article 1712: 402994 - For first time, tax question on ballot - Tallmadge Township voters will decide on 1-mill operating levy in May 3 election


==> article 1713: 403028 - Home show says spring while winter still swirls - Landscapers clear snow, then set up exhibits


==> article 1714: 402808 - Buchanan, Dean to spar at gardens - Breakfast offers left vs. right


==> article 1715: 402861 - Teens' photos get man 3 to 10 years - His family encouraged long sentence


==> article 1716: 402910 - Speed limit drops on Richmond Street - Oakleigh Avenue may be next to slow down


==> article 1717: 402982 - Tormala attacks closed sessions - Commissioner says there is more secrecy in city government despite mayor's efforts


==> article 1718: 402917 - Paws With a Cause sues over name - Agency claims business with similar name is confusing donors


==> article 1719: 403088 - Tormala attacks closed sessions - Commissioner says there is more secrecy in city government despite mayor's efforts



==> article 1758: 403787 - Sophomore knocks out Chix - Jenison's Tanis scores 33 in ot win over Zeeland


==> article 1759: 403580 - Diving may decide O-K Red Lakeshore showdown - Top-rated teams in state will compete in championship meet


==> article 1760: 403325 - Winter came and stayed too long - I once longed for gently falling snow, but now, I am ready for spring


==> article 1761: 403539 - Time for a new jail?


==> article 1762: 403231 - Allegan asks voters to replace county jail - Millage also would go to youth facility, sheriff's offices, general fund


==> article 1763: 403610 - House panel OKs road boost - Senate wants more, which Bush may veto


==> article 1764: 403377 - Hope teacher, pupil win state honors


==> article 1765: 403498 - Road work ahead?


==> article 1766: 404113 - Awards recognize top ads


==> article 1767: 404002 - Farmers see energy source blowing in the wind - MSU provides Michigan farmers with resources to explore wind energy


==> article 1768: 4



==> article 1803: 404521 - Delphi finances, leadership in question - CFO forced out after probe finds $261 million in improper accounting


==> article 1804: 404326 - Digitizing paper company bought


==> article 1805: 404406 - Students hold mini-Iditarod


==> article 1806: 404495 - Attitude, record to determine boy's fate - Police to give prosecutors report Monday on teen who threatened teachers


==> article 1807: 404466 - Agent finds knockoffs at mall - Fake Rolex watches, designer purses confiscated at 28th St. Mega mall


==> article 1808: 404381 - Hoekstra faces skeptics on Social Security overhaul


==> article 1809: 404367 - Health-care disparities to be discussed - Race, income, education and where they live often determine level of care


==> article 1810: 404477 - Cost to create regional airport studied - Tulip City Airport has been running in black for 10 years


==> article 1811: 404245 - Admitted killer puts off jail - Sudden change, uncertain timetable devastate famil



==> article 1840: 405031 - 40 assets are building blocks to better family life, student survey says - Annual town meeting focuses on common-sense ideas


==> article 1841: 404608 - Animal hospital separates the cats from the dogs - Making pets more comfortable is goal of clinic's chief veterinarian


==> article 1842: 404619 - Panel wants wider road shoulders extended - Parks Commission, willing to put money where desire is, commits $15,000


==> article 1843: 404606 - Preparations for hosting NCAA finals going swimmingly - Hope puts on Division III championships


==> article 1844: 405064 - MIAA title bright spot for women - Hope missed out on national tourney


==> article 1845: 405078 - High hopes, poor results - Big expectations dashed early on


==> article 1846: 404868 - 200 game earns Zeeland woman a cruise - Husband has to pay up on bet made 20 years ago


==> article 1847: 404777 - Will religion be wedge issue in '06? - U.S. Supreme Court ruling this summer could have major 



==> article 1884: 405507 - Donations accepted


==> article 1885: 405622 - Jammin' for Jammy


==> article 1886: 405826 - Killer's possible release sparks anger - Victim's sister says man who killed woman with ax should not go free


==> article 1887: 405783 - Corner of dreams - Developers tout their ideas for downtown 'gateway'


==> article 1888: 405592 - Race decided not by speed, but by logo - Even official who disqualified swimmer for extra Speedo insignia dislikes the rule


==> article 1889: 405676 - Big Red goes up for grabs - Holland Harbor Lighthouse Historic Commission plans to take over Holland lighthouse


==> article 1890: 405774 - Building blocks


==> article 1891: 405613 - Town meeting focuses on helping children succeed


==> article 1892: 405820 - Making up $25,000


==> article 1893: 405526 - Local giving down in wake of tsunami - Community action house asks 300 donors to give $100 each


==> article 1894: 405698 - Career Academies


==> article 1895: 405559 - All



==> article 1931: 405969 - Cities get Irish spirit, names


==> article 1932: 406112 - Panthers sharpen minds for cheer finals - Coach works on team's attitude for Saturday


==> article 1933: 406102 - Swimmer keeps focus in glare of local spotlight


==> article 1934: 405868 - A role for everyone - Maroons played as a team many dream to be part of


==> article 1935: 406809 - Resale clothing shop caters to teens, young adults - Plato's Closet features gently used apparel


==> article 1936: 406500 - Scared into bankruptcy - Threat of stricter rules forces many debtors to file now


==> article 1937: 406735 - Steelcase helps Boeing's plant fly - Office-furniture maker's product transforms 737 facility


==> article 1938: 406377 - Coleman recalls furnaces


==> article 1939: 406432 - Current woes at GM and Ford can be traced to Roger Smith - Former chairman's missteps have stuck automakers with costly labor
pacts



==> article 1940: 406833 - Writer gets his chance to speak - Racial s



==> article 1978: 406939 - Lot needs appraisal before sale to B.O.B. - Commission wants money's worth for parking lot that could become theater, hotel


==> article 1979: 407102 - Hearing recommends outpatient surgery center - Seven other projects approved, but lower prices recommended


==> article 1980: 407058 - New air rules greeted with favor in Michigan - EPA's regional approach looks to reduce pollution that drifts across borders


==> article 1981: 407266 - Not quite ripe? - Planners set hearing for shopping center on old orchard


==> article 1982: 407276 - New air rules greeted with favor in Michigan - EPA's regional approach looks to reduce pollution that drifts across borders


==> article 1983: 407205 - Music to inspire dreams - Latino students celebrate goals


==> article 1984: 407110 - Road will be widened for South Beltline traffic


==> article 1985: 407298 - With friends like these ... - 2006 elections are already in mind, partisan divisions intensify in Lansing


=



==> article 2023: 407806 - Boy hurt when sled rams into garage - 12-year-old was being pulled behind an all-terrain vehicle


==> article 2024: 407658 - Tribute band salutes teacher - Past members of high school band return to celebrate director's 25th year


==> article 2025: 407693 - Will youth center be big bully? - Salvation Army says big-bucks plan not out to dwarf other programs


==> article 2026: 408067 - Spring fever, Irish style - Snowy weather adds a little white with green of St. Patrick's parade


==> article 2027: 407615 - McCain says visit not campaign stop - Appearances Monday in West Michigan are to do "a little party stuff"
ValueError parsing OpenCalais JSON for Article 407615 - raw response body: You exceeded the concurrent request limit for your license key. Please try again later or contact support to upgrade your license.
======> In code_article_data(): ERROR - Error: ValueError parsing OpenCalais JSON for Article 407615 - raw response body: You exceeded the con



==> article 2061: 408449 - Head of the class - Hope excels as host of Division III national swim meet


==> article 2062: 408564 - Firm picks area for expansion - Emerald Graphics will begin production in Kentwood, Cascade Township


==> article 2063: 408581 - Retail, apartments planned for vacant buildings


==> article 2064: 408752 - Firm picks area for expansion - Emerald Graphics will begin production in Kentwood, Cascade Township


==> article 2065: 408662 - Smoke signals - Company's ban on smokers gains acceptance, attorney says


==> article 2066: 408474 - Employers expect to hire, not lay off, this spring - Survey shows stronger hiring intentions between April and June


==> article 2067: 408685 - Alticor urges better workers' health


==> article 2068: 408670 - CEO says Delphi fiscally sound - Accounting woes not reason for his departure, Battenberg says


==> article 2069: 408538 - Kimball ready to talk cuts in budget - City manager will tell commissioners they need to trim



==> article 2106: 409258 - Minister's message: Be smart - He must live with consequences of earlier, carefree lifestyle


==> article 2107: 408846 - Up for bids


==> article 2108: 409044 - Communication problems slowed pileup rescue - Authorities say they would establish common frequencies for disasters


==> article 2109: 409130 - Police try to revive Child Watch program


==> article 2110: 409124 - Classroom pledge heads to House - Senate unanimously passes bill to allow for recitation in schools


==> article 2111: 409224 - Students, parents get fair look at colleges


==> article 2112: 409215 - Advocate for disabled admits to defrauding investors - 'I did not use the money as promised,' he says as he pleads guilty
to scam



==> article 2113: 408965 - Commissioners float ideas to plug budget shortfall - Grand Rapids faces an $11 million deficit, and some officials look to
Lansing for help



==> article 2114: 409118 - Cabbies to have drug tests when applying for licenses - New r



==> article 2152: 409759 - Baby girl was not sold, says Michigan woman - Lake County couple were taking legal steps for full guardianship of Kentucky
infant, lawyer says



==> article 2153: 409812 - Baby girl was not sold, says Michigan woman - Lake County couple were taking legal steps for full guardianship of Kentucky
infant, lawyer says



==> article 2154: 409979 - From schools to sports bars, we celebrate our inner Irish


==> article 2155: 409909 - Farmland sought for conservation effort - State program encourages improvements at farms to protect water quality


==> article 2156: 409649 - Developer tweaks size, price for Baker Lofts - Condominiums also will be for sale sooner


==> article 2157: 409575 - Big City Swing - Holland Christian group presents variety of musical styles at showcase


==> article 2158: 409945 - Driver to stand trial for girl's crash death - Defense says child died because she did not have seat belt


==> article 2159: 409764 - Church services at park i



==> article 2197: 410298 - Lords of the pool - Holland native helps Kenyon college lead swimming finals


==> article 2198: 410223 - Sprague hits milestone with NASCAR Trucks - Tonight he will be second driver to reach 200 races


==> article 2199: 410037 - Relay gives Panthers fast start at finals - Winning 200 medley was boost, since Chix won race at earlier swim meet


==> article 2200: 410076 - Church offers free help for tax time - Pastor says people with financial problems who seek help do not have to attend services


==> article 2201: 410110 - Closing off beaches hides God's gift - Walks along beach shows connections to nature


==> article 2202: 410331 - Competitive, but still too high - State insurance chief seeks legal tools to protect consumers


==> article 2203: 410356 - Village seeks payment from gas station owner - Middleville says property is contaminated


==> article 2204: 410236 - Allegan County won't limit length of board chairman's tenure - Proposal would have r



==> article 2243: 410756 - Diekema siblings wear record smiles after meet finishes - Jordan (Zeeland), Megan (Calvin) earn honors


==> article 2244: 410913 - Little time to celebrate for Heyboe - Holland Christian champ heads to series in Cleveland


==> article 2245: 410952 - A luge official on solid ground - Holland's Johnson back from officiating championships


==> article 2246: 411009 - Solid Steele, but she wanted more - Hope senior sought stronger final showing at D-III final


==> article 2247: 411085 - 'I've been dead for over 200 years' - Director brings classical composers to life


==> article 2248: 411008 - Democrats slam plans for Social Security - They critique Bush's call for adding private accounts to retirement program
ValueError parsing OpenCalais JSON for Article 411008 - raw response body: You exceeded the concurrent request limit for your license key. Please try again later or contact support to upgrade your license.
======> In code_article_data(): ERROR - Erro

- started: execution queued 22:42:01 2019-07-31


# TODO

- Back to [Table of Contents](#Table-of-Contents)

TODO:

- Save log of coding first 4990 of next round of data.
- for next round of coding, sort on publication date, descending, so we fill in the year before and after the layoffs first.
- start loading data from XML
- move data from Article_Data into context.
- make the network data creator work against context, then generalize it for tie and node types.
- adjust django logging to output DEBUG, then test Article.filter_articles() to see where QuerySet is evaluated (DISTINCT check?).
- think how we specify which class to use for author strings - needs to be speced to an interface, but not just a newsbank one - so, abstraction here should be higher up - in shared?
